clients/drcachesim/tracer/kcore_copy.cpp

```diff
@@ -0,0 +1,433 @@
+ * Copyright (c) 2022 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include <elf.h>
+#include <inttypes.h>
+#include <string.h>
+#include <stdio.h>
+#include <fstream>
+
+#include "../common/utils.h"
+#include "dr_api.h"
+#include "kimage.h"
+#include "kcore_copy.h"
+
+#define KIMAGE_FILE_NAME "kimage"
+#define MODULES_FILE_NAME "modules"
+#define MODULES_FILE_PATH "/proc/" MODULES_FILE_NAME
+#define KALLSYMS_FILE_NAME "kallsyms"
+#define KALLSYMS_FILE_PATH "/proc/" KALLSYMS_FILE_NAME
+#define KCORE_FILE_NAME "kcore"
+#define KCORE_FILE_PATH "/proc/" KCORE_FILE_NAME
+
+#define KERNEL_SYMBOL_MAX_LEN 300
+
+/* This struct type defines the information of every module read from /proc/module.
+ * We store all information on a linked list.
+ */
+struct proc_module_t {
+    proc_module_t *next;
+
+    /* The start virtual address of the module in memory. */
+    uint64_t start;
+
+    /* The end virtual address of the module in memory. */
+    uint64_t end;
+};
+
+/* This struct type defines the metadata of every code segment read from  /proc/kcore. */
+struct proc_kcore_code_segment_t {
+    /* The start offset of the code segment in /proc/kcore. */
+    uint64_t start;
+
+    /* The length of the code segment. */
+    ssize_t len;
+
+    /* The virtual address of the code segment in memory. */
+    uint64_t vaddr;
+
+    /* The buffer to store the code segment. */
+    char *buf;
```

Comment:
Do we need this intermediate copy?  How much data are we talking about?  If it is a lot of data it would be better to just record the offsets and only read it once, in pieces, when we write it out to the output file.

Reply:
I will add an option on the next PR to let the user specify whether to dump kernel binary.